pageserver/src/walreceiver/connection_manager.rs

```diff
@@ -0,0 +1,1134 @@
+                    new_candidate.safekeeper_id,
+                    new_candidate.wal_producer_connstr,
+                )
+                .await
+        }
+    }
+}
+
+/// Endlessly try to subscribe for broker updates for a given timeline.
+/// If there are no safekeepers to maintain the lease, the timeline subscription will be unavailable in the broker and the operation will fail constantly.
+/// This is ok, pageservers should anyway try subscribing (with some backoff) since it's the only way they can get the timeline WAL anyway.
+async fn subscribe_for_timeline_updates(
+    etcd_client: &mut Client,
+    broker_prefix: &str,
+    id: ZTenantTimelineId,
+) -> BrokerSubscription<SkTimelineInfo> {
+    let mut attempt = 0;
+    loop {
+        exponential_backoff(
+            attempt,
+            DEFAULT_BASE_BACKOFF_SECONDS,
+            DEFAULT_MAX_BACKOFF_SECONDS,
+        )
+        .await;
+        attempt += 1;
+
+        match etcd_broker::subscribe_for_json_values(
+            etcd_client,
+            SubscriptionKey::sk_timeline_info(broker_prefix.to_owned(), id),
+        )
+        .instrument(info_span!("etcd_subscription"))
+        .await
+        {
+            Ok(new_subscription) => {
+                return new_subscription;
+            }
+            Err(e) => {
+                warn!("Attempt #{attempt}, failed to subscribe for timeline {id} updates in etcd: {e:#}");
+                continue;
+            }
+        }
+    }
+}
+
+const DEFAULT_BASE_BACKOFF_SECONDS: f64 = 2.0;
+const DEFAULT_MAX_BACKOFF_SECONDS: f64 = 60.0;
+
+async fn exponential_backoff(n: u32, base: f64, max_seconds: f64) {
+    if n == 0 {
+        return;
+    }
+    let seconds_to_wait = base.powf(f64::from(n) - 1.0).min(max_seconds);
+    info!("Backoff: waiting {seconds_to_wait} seconds before proceeding with the task");
+    tokio::time::sleep(Duration::from_secs_f64(seconds_to_wait)).await;
+}
+
+/// All data that's needed to run endless broker loop and keep the WAL streaming connection alive, if possible.
+struct WalreceiverState {
+    id: ZTenantTimelineId,
+    /// Use pageserver data about the timeline to filter out some of the safekeepers.
+    local_timeline: Arc<DatadirTimelineImpl>,
+    /// The timeout on the connection to safekeeper for WAL streaming.
+    wal_connect_timeout: Duration,
+    /// The timeout to use to determine when the current connection is "stale" and reconnect to the other one.
+    lagging_wal_timeout: Duration,
+    /// The Lsn lag to use to determine when the current connection is lagging to much behind and reconnect to the other one.
+    max_lsn_wal_lag: NonZeroU64,
+    /// Current connection to safekeeper for WAL streaming.
+    wal_connection: Option<WalConnection>,
+    wal_connection_attempts: HashMap<NodeId, u32>,
+    /// Data about all timelines, available for connection, fetched from etcd.
+    wal_stream_candidates: HashMap<SubscriptionFullKey, EtcdSkTimeline>,
```

Comment:
Why key here is `SubscriptionFullKey`, won't just safekeeper node id be simpler?